calyx/src/passes/group_to_seq.rs

```diff
@@ -0,0 +1,217 @@
+use std::rc::Rc;
+
+#[derive(Default)]
+/// Transforms a group into a seq of smaller groups, if possible.
+/// Currently, in order for a group to be transformed, must a) consist of only
+/// writes to non-combination cells or the group's done port, and b) there must be a clear,linear ordering of
+/// the execution of each cell in the group by looking at go-done assignments and c) group[done] = cell.done for
+/// some cell.
+pub struct GroupToSeq;
+
+impl Named for GroupToSeq {
+    fn name() -> &'static str {
+        "group-to-seq"
+    }
+
+    fn description() -> &'static str {
+        "split groups under correct conditions"
+    }
+}
+
+//If asmt is a write to a cell named name returns Some(name).
+//If asmt is a write to a group port, returns None.
+fn is_write_to(asmt: &ir::Assignment) -> Option<ir::Id> {
+    match &asmt.dst.borrow().parent {
+        ir::PortParent::Cell(cell) => {
+            Some(cell.upgrade().borrow().name().clone())
+        }
+        ir::PortParent::Group(_) => None,
+    }
+}
+
+//Given asmt a.go = b.done, return (a1, a2), where a1 is group[done] = b.done,
+//and a2 is a.go = 1'd1.
+fn split_go_done(
+    builder: &mut ir::Builder,
+    asmt: ir::Assignment,
+    group: ir::WRC<ir::Group>,
+) -> (ir::Assignment, ir::Assignment) {
+    let con = builder.add_constant(1, 1);
+    let src = ir::Port {
+        name: ir::Id::new("const 1", None),
+        width: 1,
+        direction: ir::Direction::Output,
+        parent: ir::PortParent::Cell(ir::WRC::from(&con)),
+        attributes: ir::Attributes::default(),
+    };
+    let src_ref = Rc::new(RefCell::new(src));
+
+    let dst = ir::Port {
+        name: ir::Id::new("done", None),
+        width: 1,
+        direction: ir::Direction::Input,
+        parent: ir::PortParent::Group(group),
+        attributes: ir::Attributes::default(),
+    };
+    let dst_ref = Rc::new(RefCell::new(dst));
+    (
+        builder.build_assignment(dst_ref, asmt.src, ir::Guard::True),
+        builder.build_assignment(asmt.dst, src_ref, ir::Guard::True),
+    )
+}
+
+//Given asmt old_group[done] = guard? a.done, return group[done] = guard? a.done.
+fn rename_group_done(
+    builder: &mut ir::Builder,
+    asmt: ir::Assignment,
+    group: ir::WRC<ir::Group>,
+) -> ir::Assignment {
+    let dst = ir::Port {
+        name: ir::Id::new("done", None),
+        width: 1,
+        direction: ir::Direction::Input,
+        parent: ir::PortParent::Group(group),
+        attributes: ir::Attributes::default(),
+    };
+    let dst_ref = Rc::new(RefCell::new(dst));
+    builder.build_assignment(dst_ref, asmt.src, *asmt.guard)
+}
+
+//Gets the name of the port parent
+fn get_parent_name(port: &ir::RRC<ir::Port>) -> ir::Id {
+    match &port.borrow().parent {
+        ir::PortParent::Cell(cell) => cell.upgrade().borrow().name().clone(),
+        ir::PortParent::Group(group) => group.upgrade().borrow().name().clone(),
+    }
+}
+
+impl Visitor for GroupToSeq {
+    fn enable(
```

Comment:
Use the `Visitor::start` method to iterate over `comp.groups` directly instead of using `enable`. This saves having to traverse the control tree. Also, make this function return `Action::Stop` so that the control tree doesn't traversed.